File: 2020/day-3/part-1/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

func readInput(filepath string) [][]string {
	file, err := os.Open(filepath)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	input := make([][]string, 0)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		characters := strings.Split(line, "")
		input = append(input, characters)
	}
	return input
}

func calculate(input [][]string, right int, down int) int {
	width := len(input[0])
	height := len(input)
	x, y := 0, 0
	count := 0
	for x < height {
		if y > width - 1 {
			y = y - width
		}
		if input[x][y] == string('#') {
			count += 1
		}
		x = x + down
		y = y + right
	}
	return count
}

func main() {
	down, right := 1, 3
	input := readInput("input.txt")
	result := calculate(input, right, down)
	fmt.Println(result)
}

File: 2020/day-3/part-1/main.py
#!/usr/bin/env python3

def read_input(filepath):
    input = []
    with open(filepath) as fd:
        lines = fd.readlines()
        for line in lines:
            line = line.strip()
            char_arr = []
            for character in line:
                char_arr.append(character)
            input.append(char_arr)
    return input

def calculate(input, right, down):
    width = len(input[0])
    height = len(input)
    x, y = 0, 0
    count = 0
    while x < height:
        if y > width - 1:
            y = y - width
        if input[x][y] == '#':
            count += 1
        x = x + down
        y = y + right
    return count

def main():
    down = 1
    right = 3
    input = read_input("input.txt")
    result = calculate(input, right, down)
    print(result)

if __name__ == '__main__':
    main()